### DISCUSSION TITLE
Permanently running background task that listens to Postgres notifications and sends data to websocket

### DISCUSSION THREAD

**Discussion by artemonsh** [fastapi/fastapi#6221] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
import aiopg
from fastapi import FastAPI, WebSocket


dsn = "dbname=aiopg user=aiopg password=passwd host=127.0.0.1"
app = FastAPI()


class ConnectionManager:
    self.count_connections = 0
    # other class functions and variables are taken from FastAPI docs
    ...


manager = ConnectionManager()


async def send_and_receive_data(websocket: WebSocket):
    data = await websocket.receive_json()
    await websocket.send_text('Thanks for the message')
    # then process received data


# taken from official aiopg documentation
# the function listens to PostgreSQL notifications
async def listen(conn):
    async with conn.cursor() as cur:
        await cur.execute("LISTEN channel")
        while True:
            msg = await conn.notifies.get()


async def postgres_listen():
    async with aiopg.connect(dsn) as listenConn:
        listener = listen(listenConn)
        await listener


@app.get("/")
def read_root():
    return {"Hello": "World"}


@app.websocket("/")
async def websocket_endpoint(websocket: WebSocket):
    await manager.connect(websocket)
    manager.count_connections += 1

    if manager.count_connections == 1:
        await asyncio.gather(
            send_and_receive_data(websocket),
            postgres_listen()
        )
    else:
        await send_and_receive_data(websocket)
```

### Description

I am building an app with Vue.js, FastAPI and PostgreSQL. In this example I attempt to use listen/notify from Postgres and implement it in the websocket. I also use a lot of usual http endpoints along with the websocket endpoint.

I want to run a permanent background asynchronous function at the start of the FastAPI app that will then send messages to all websocket clients/connections. So, when I use `uvicorn  main:app` it should not only run the FastAPI app but also my background function `postgres_listen()`, which notifies all websocket users, when a new row is added to the table in the database.

I know that I can use `asyncio.create_task()` and place it in the `on_*` event, or even place it after the `manager = ConnectionManager()` row, but it will not work in my case! Because after any http request (for instance, `read_root()` function), I will get the same error described below.

You see that I use a strange way to run my `postgres_listen()` function in my `websocket_endpoint()` function only when the first client connects to the websocket. Any subsequent client connection does not run/trigger this function again. And everything works fine... until the first client/user disconnects (for example, closes browser tab). When it happens, I immediately get the `GeneratorExit` error caused by `psycopg2.OperationalError`: 
```
Future exception was never retrieved
future: <Future finished exception=OperationalError('Connection closed')>
psycopg2.OperationalError: Connection closed
Task was destroyed but it is pending!
task: <Task pending name='Task-18' coro=<Queue.get() done, defined at 
/home/user/anaconda3/lib/python3.8/asyncio/queues.py:154> wait_for=<Future cancelled>>
```
The error comes from the `listen()` function. After this error, I will not get any notification from the database as the asyncio's `Task` is cancelled. There is nothing wrong with the `psycopg2`, `aiopg` or `asyncio`. The problem is that I don't understand where to put the `postgres_listen()` function so it will not be cancelled after the first client disconnects. From my understanding, I can easily write a python script that will connect to the websocket (so I will be the first client of the websocket) and then run forever so I will not get the `psycopg2.OperationalError` exception again and the `Task` won't be cancelled, but it does not seem right to do so.

My question is: where should I put `postgres_listen()` function, so the first connection to websocket may be disconnected with no consequences?

P.S. `asyncio.shield()` also does not work

### Operating System

Linux

### FastAPI Version

0.77.1

### Python Version

3.8.5

**Comment by JarroVGIT** [comment]
I would recommend to use Broadcaster, that was specifically built for the use case that you are doing. 

https://pypi.org/project/broadcaster/

Although this is an alpha piece of software, it might give you indicators on how to achieve what you want to achieve :)

**Comment by JarroVGIT** [comment]
Alright, I was curious if I could hack something together that would help you out a bit. See the following code:

```python
import queue
from typing import Any
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from asyncio import Queue, Task
import asyncio

import uvicorn

class Listener:
    def __init__(self):
        self.subscribers: list[Queue] = []
        self.listener_task: Task
    async def subscribe(self, q: Queue):
        self.subscribers.append(q)

    #this function is for testing, it is called when a certain endpoint is called
    #and is instead of the whole start_listening logic.
    async def receive_and_publish_message(self, msg: Any):
        for q in self.subscribers:
            try:
                q.put_nowait(str(msg))
            except Exception as e:
                raise e

    async def start_listening(self, conn):
        self.listener_task = asyncio.create_task(self._listener(conn=conn))


    async def _listener(self, conn) -> None:
        async with conn.cursor() as cur:
            await cur.execute("LISTEN channel")
            while True:
                msg = await conn.notifies.get()
                for q in self.subscribers:
                    await q.put(msg)

    async def stop_listening(self):
        if self.listener_task.done():
            self.listener_task.result()
        else:
            self.listener_task.cancel()


global_listener = Listener()

app = FastAPI()

@app.on_event("startup")
async def startup_event():
    # normally, you would call the start_listener
    # conn = some_method_to_get_postgres_connection()
    # global_listener.start_listening(conn=conn)
    # as i do not have a stream listener, i am passing.
    pass

@app.on_event("shutdown")
async def shutdown_event():
    # Here you would normally call the stop_listener method
    pass


@app.get('/add_item/{item}')
async def add_item(item: str):
    await global_listener.receive_and_publish_message(item)
    return {"published_message:": item}

@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    q: asyncio.Queue = asyncio.Queue()
    await global_listener.subscribe(q=q)
    try:
        while True:
            data = await q.get()
            await websocket.send_text(data)
    except WebSocketDisconnect:
            pass

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

This would run as is, and new joiners on your websocket would start seeing the new messages as well. I actually tried that:
<img width="872" alt="image" src="https://user-images.githubusercontent.com/13659033/172928437-17029b8d-ed17-425f-ba34-818d83481b75.png">

So, this is just to get you going and is very much NOT a copy paste for your code base. For example, I didn't have any 'realtime listener' available so I created a non-realtime version of it (method `receive_and_publish_message()`). That method is called whenever I made a GET request to `/add_item/{item}`, but the concepts are relatively the same. Note that his is incredibly inspired by the codebase of Broadcaster, which I mentioned in my previous comment. 

The basics are as follows:
- You need a handler class for your global listener (in your case, the global listener is `msg = await conn.notifies.get()`
- New connections to your websocket need to be able to subscribe to that. 
- Whenever a new message comes in, into your global listener, it needs to be fed into all the websocket connections. 

I provided some code (which I haven't tested, so be aware) for you to tinker around with and provided some comments to illustrate how it would fit in your use case. I hope this helps, if you have questions just let me know :)

**Comment by JarroVGIT** [comment, marked as answer]
Okay this kept bugging me, so I've created a more relevant example for you. The code can be found here:

https://github.com/JarroVGIT/fastapi-github-issues/tree/master/5015

The app.py is basically the same as above, but with one change: it will listen to a websocket and publishes all incoming messages to all connections to the websockets in app.py.

The generator.py is a simple FastAPI app that has a websocket (that our example above listens to) that emits a message every 2 seconds to every connection it gets.

To try this out:
1. Start generator.py (e.g. `python3 generator.py` on your command line when in your working folder)
2. Start app.py (either debug mode in VScode or same as above)
3. Listen to `http://localhost:8000/ws` (= endpoint in app.py) with several clients, you will see that they will all join in the same message streak.

Let me know if this helps you out!

**Comment by adriangb** [comment]
I think this should work pretty well for your use case:

```python
@app.startup
async def startup() -> None:
    listener = Listener()
    listener.connect()
    app.dependency_overrides[Listener] = lambda: listener

@app.get("/"
async def endpoint(listener: Listener = Depends()):
    ....
```

I'd be cautious of having a global `listener` object that has non-thread/process safe stuff (e.g. a reference to an asyncio task) since that could get wonky if used with process forking.

**Comment by artemonsh** [comment]
@JarroVGIT, huge respect for such well-documented code and reproducible example as well. It works as expected, it works like magic.
